internal/authorization/rbac_translator.go

```diff
@@ -0,0 +1,361 @@
+		if !ok {
+			return nil, fmt.Errorf("\"headers\" %d: \"key\" %v is not a string", i, key)
+		}
+		// TODO(ashithasantosh): Return error for unsupported headers- "hop-by-hop",
+		// pseudo headers, "Host" header and headers prefixed with "grpc-".
+		values, found := h["values"]
+		if !found {
+			return nil, fmt.Errorf("\"headers\" %d: \"values\" is not present", i)
+		}
+		or, err := parseHeaderValues(keyStr, values)
+		if err != nil {
+			return nil, fmt.Errorf("\"headers\" %d: %v", i, err)
+		}
+		and = append(and, permissionOr(or))
+	}
+	return and, nil
+}
+
+func parseRequest(request interface{}) (*v3rbacpb.Permission, error) {
+	var and []*v3rbacpb.Permission
+	r, ok := request.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("\"request\" is not an object")
+	}
+	if p, found := r["paths"]; found {
+		or, err := parsePaths(p)
+		if err != nil {
+			return nil, err
+		}
+		if len(or) > 0 {
+			and = append(and, permissionOr(or))
+		}
+	}
+	if h, found := r["headers"]; found {
+		subAnd, err := parseHeaders(h)
+		if err != nil {
+			return nil, err
+		}
+		if len(subAnd) > 0 {
+			and = append(and, permissionAnd(subAnd))
+		}
+	}
+	if len(and) > 0 {
+		return permissionAnd(and), nil
+	}
+	return permissionAny(), nil
+}
+
+func parseRulesArray(rules interface{}, prefixName string) (map[string]*v3rbacpb.Policy, error) {
+	rulesArray, ok := rules.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("rules is not an array")
+	}
+	policies := make(map[string]*v3rbacpb.Policy)
+	for i, r := range rulesArray {
+		rule := r.(map[string]interface{})
+		name, found := rule["name"]
+		if !found {
+			return policies, fmt.Errorf("%d: \"name\" is not present", i)
+		}
+		nameStr, ok := name.(string)
+		if !ok {
+			return policies, fmt.Errorf("%d: \"name\" %v is not a string", i, name)
+		}
+		nameStr = prefixName + "_" + nameStr
+		var principals []*v3rbacpb.Principal
+		if source, found := rule["source"]; found {
+			principal, err := parsePeer(source)
+			if err != nil {
+				return nil, fmt.Errorf("%d: %v", i, err)
+			}
+			principals = append(principals, principal)
+		} else {
+			principals = append(principals, principalAny())
+		}
+		var permissions []*v3rbacpb.Permission
+		if request, found := rule["request"]; found {
+			permission, err := parseRequest(request)
+			if err != nil {
+				return nil, fmt.Errorf("%d: %v", i, err)
+			}
+			permissions = append(permissions, permission)
+		} else {
+			permissions = append(permissions, permissionAny())
+		}
+		policies[nameStr] = &v3rbacpb.Policy{
+			Principals:  principals,
+			Permissions: permissions,
+		}
+	}
+	return policies, nil
+}
+
+// Translates a gRPC authorization policy in JSON string to two RBAC policies, a deny
+// RBAC policy followed by an allow RBAC policy. If the policy cannot be parsed or is
+// invalid, an error will be returned.
+func translatePolicy(policy string) (*v3rbacpb.RBAC, *v3rbacpb.RBAC, error) {
+	var jsonData map[string]interface{}
```

Comment:
Instead of unmarshalling into a `map`, why not define a `struct` that represents the SDK API's schema?